legion/engine/physics/physx/physx_integration_helpers.cpp

```diff
@@ -0,0 +1,85 @@
+        staticCollider->userData = ent.data;
+
+        wrapper.physicsActor = staticCollider;
+        sceneInfo.scene->addActor(*staticCollider);
+    }
+
+    template<class PxGeometry, class ...GeometryArgs>
+    void instantiateDynamicActorWith(physx::PxPhysics* sdk, PhysxInternalWrapper& wrapper,
+        const physx::PxTransform& globalTransform, const physx::PxTransform& localTransform, const PhysxEnviromentInfo& sceneInfo, ecs::entity ent, GeometryArgs... geometryArgs)
+    {
+        PxRigidDynamic* rb = PxCreateDynamic(*getSDK(), globalTransform,
+            PxGeometry(std::forward<GeometryArgs>(geometryArgs)...), *sceneInfo.defaultMaterial, 1.0f, localTransform);
+        rb->userData = ent.data;
+
+        wrapper.physicsActor = rb;
+        sceneInfo.scene->addActor(*rb);
+    }
+
+
+    template<class PxGeometry, class ...GeometryArgs>
+    void instantiateNextCollider(physx::PxPhysics* sdk, PhysxInternalWrapper& wrapper, const physx::PxTransform& globalTransform, const physx::PxTransform& localTransform,
+        const PhysxEnviromentInfo& sceneInfo, GeometryArgs ...geometryArgs)
+    {
+        PxShape* shape = getSDK()->createShape(PxGeometry(std::forward<GeometryArgs>(geometryArgs)...), *sceneInfo.defaultMaterial, true);
+
+        PxRigidActor* rigid = static_cast<PxRigidActor*>(wrapper.physicsActor);
+
+        shape->setLocalPose(localTransform);
+        rigid->attachShape(*shape);
+        shape->release();
+    }
+
+#define DECLARE_STATIC_COLLIDER_TEMPLATE_INSTANTIATION(PxGeometry,...)  template void instantiateStaticActorWith<PxGeometry, __VA_ARGS__>(physx::PxPhysics* sdk, PhysxInternalWrapper& wrapper, const physx::PxTransform& outGlobalTransform, const physx::PxTransform& outLocalTransform, const PhysxEnviromentInfo& sceneInfo, ecs::entity ent, __VA_ARGS__)
+
+    DECLARE_STATIC_COLLIDER_TEMPLATE_INSTANTIATION(PxSphereGeometry, float);
+    DECLARE_STATIC_COLLIDER_TEMPLATE_INSTANTIATION(PxBoxGeometry, const PxVec3&);
+
+#define DECLARE_DYNAMIC_COLLIDER_TEMPLATE_INSTANTIATION(PxGeometry,...) template void instantiateDynamicActorWith<PxGeometry, __VA_ARGS__>(physx::PxPhysics* sdk, PhysxInternalWrapper& wrapper,const physx::PxTransform& outGlobalTransform, const physx::PxTransform& outLocalTransform, const PhysxEnviromentInfo& sceneInfo, ecs::entity ent, __VA_ARGS__);
+
+    DECLARE_DYNAMIC_COLLIDER_TEMPLATE_INSTANTIATION(PxSphereGeometry, float);
+    DECLARE_DYNAMIC_COLLIDER_TEMPLATE_INSTANTIATION(PxBoxGeometry, const PxVec3&);
+
+#define DECLARE_NEXT_COLLIDER_TEMPLATE_INSTANTIATION(PxGeometry,...) template void instantiateNextCollider<PxGeometry, __VA_ARGS__>(physx::PxPhysics* sdk, PhysxInternalWrapper& wrapper, const physx::PxTransform& outGlobalTransform, const physx::PxTransform& outLocalTransform, const PhysxEnviromentInfo& sceneInfo, __VA_ARGS__);
+
+    DECLARE_NEXT_COLLIDER_TEMPLATE_INSTANTIATION(PxSphereGeometry, float);
+    DECLARE_NEXT_COLLIDER_TEMPLATE_INSTANTIATION(PxBoxGeometry, const PxVec3&);
```

Comment:
wait why do we do this? (also missing ; on line 70)

Reply:
I don't want to include `<physx/PxPhysicsAPI.h> ` in the hpp file. Which means the definition of the templated function instantiateDynamicActor and others need to be in the cpp file , since they are templated I need to do explicit template instantiation.